src/main/kotlin/Main.kt

```diff
@@ -96,6 +106,99 @@ private fun executeTests(tests: Collection<Test>): Iterable<ResultContext> {
     return fullResults
 }
 
+private fun generateTests(): Collection<Test> {
+    val allRelations = ProofSearcher().addAllProofs().findNewRelations(RelationLoader.relations)
+    //ProofLimiter(3).limit(allRelations)
+    return TestGenerator().addAllTests().generateTests(allRelations)
+}
+
+
+val OPERATORS = listOf("||", "\\\\", "&&", "consistency:", "refinement:")
+private fun sortTests(engine: EngineConfiguration, tests: Collection<Test>) : Collection<Test> {
+    var out = ArrayList(tests)
+
+    if (engine.queryComplexity != null) { //Query Complexity
+        val upper: Int; val lower: Int
+        if (engine.queryComplexity.size >= 2) {
+            upper = engine.queryComplexity[1]
+            lower = engine.queryComplexity.first()
+        } else {
+            upper = engine.queryComplexity.firstOrNull() ?: Int.MAX_VALUE
+            lower = 0
+        }
+
+        if (upper < lower)
+            throw Exception("The upper bound for `queryComplexity` can't be less than the lower bound")
+
+        out = ArrayList(out.filter { x ->
+            x.queries().all { y ->
+                var c = 0
+                for (op in OPERATORS) {
+                    c += y.split(op)
+                        .dropLastWhile { it.isEmpty() }
+                        .toTypedArray().size - 1
+                }
+                c in lower..upper
+            }
+        })
+    }
```

Comment:
I have a hard time figuring out this code (Maybe i dont know idiomatic Kotlin), but a comment would help.